### ISSUE TITLE
Permisson problems updating "go" on Fedora 33 beta

### ISSUE THREAD

**Issue by donthomas21** [r-darwish/topgrade#542]
I have installed Fedora 33 beta on two computers, topgrade worked perfectly until I installed go using dnf

A number of updates appear to be not allowed, here are a couple examples:

go get cmd/internal/obj/arm: copying /home/dthomas/.cache/go-build/9c/9c5ecec0f072eae8d92a5fae69b944566e51e1a3ac6a1dd74744805f81c034e6-d: open /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/arm.a: permission denied 
go get cmd/internal/obj/arm64: copying /home/dthomas/.cache/go-build/3b/3b1c459c7534e8a25984a8b2ffbd4658b3d9f7bd8e6072da60eae4ddcb465b73-d: open /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/arm64.a: permission denied
go get cmd/internal/obj/mips: copying /home/dthomas/.cache/go-build/d8/d8e3e5ed1828ebcbe7bc5eadd6e983fe6b3f51baf3585a414c53b72479f72cd7-d: open /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/mips.a: permission denied 
go get cmd/internal/obj/ppc64: copying /home/dthomas/.cache/go-build/9a/9a1c18c66649a9dc9a60adf0036cbbffcb117a3be8ff8b12a53c8f3aa24dbdbb-d: open /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/ppc64.a: permission denied 

I have installed the musl version of topgrade on Fedora 33 beta

- If a specific step is failing, what happens when you run the command on your own? (Run `topgrade --dry-run` to see which commands Topgrade is running)

Dry running: /usr/bin/go get -u all

If I run the command as is I get the same errors as running normally. If I run it with sudo the update runs perfectly.

<details><pre>
topgrade -v
 2020-10-07T09:53:26.622Z DEBUG topgrade::config > Configuration at /home/dthomas/.config/topgrade.toml
 2020-10-07T09:53:26.622Z DEBUG topgrade::config > Loaded configuration: ConfigFile { pre_commands: None, post_commands: None, commands: None, git_repos: None, predefined_git_repos: None, disable: None, ignore_failures: None, remote_topgrades: None, remote_topgrade_path: None, ssh_arguments: None, git_arguments: None, tmux_arguments: None, set_title: None, assume_yes: None, yay_arguments: None, no_retry: None, run_in_tmux: None, cleanup: None, notify_each_step: None, accept_all_windows_updates: None, only: None, composer: None, brew: None, linux: None, git: None, windows: None, vagrant: None }
 2020-10-07T09:53:26.623Z DEBUG topgrade::utils  > Detected "/usr/bin/notify-send" as "notify-send"
 2020-10-07T09:53:26.623Z DEBUG topgrade         > Version: 5.8.1
 2020-10-07T09:53:26.623Z DEBUG topgrade         > OS: x86_64-unknown-linux-musl
 2020-10-07T09:53:26.623Z DEBUG topgrade         > Args { inner: ["topgrade", "-v"] }
 2020-10-07T09:53:26.623Z DEBUG topgrade         > Binary path: Ok("/home/dthomas/bin/topgrade")
 2020-10-07T09:53:26.623Z DEBUG topgrade         > Self Update: true
 2020-10-07T09:53:26.623Z DEBUG topgrade::utils  > Detected "/usr/bin/git" as "git"
 2020-10-07T09:53:26.623Z DEBUG topgrade::utils  > Detected "/usr/bin/sudo" as "sudo"

―― 20:53:26 - Self update ――――――――――――――――――――――――――――――――――――――――――――――――――――――
Topgrade is up-to-date
 2020-10-07T09:53:26.971Z DEBUG topgrade::utils  > Cannot find "pwsh"
 2020-10-07T09:53:26.971Z DEBUG topgrade::utils  > Cannot find "powershell"
 2020-10-07T09:53:26.971Z DEBUG topgrade::runner > Step "System update"

―― 20:53:26 - System update ――――――――――――――――――――――――――――――――――――――――――――――――――――
 2020-10-07T09:53:26.972Z DEBUG topgrade::utils  > Path "/usr/bin/rpm-ostree" doesn't exist
 2020-10-07T09:53:26.972Z DEBUG topgrade::utils  > Path "/usr/bin/dnf" exists
 2020-10-07T09:53:26.972Z DEBUG topgrade::executor > Running "/usr/bin/sudo" "/usr/bin/dnf" "upgrade"
Last metadata expiration check: 1:35:14 ago on Wed 07 Oct 2020 19:18:14.
Dependencies resolved.
Nothing to do.
Complete!
 2020-10-07T09:53:31.163Z DEBUG topgrade::runner   > Step "etc-update"
 2020-10-07T09:53:31.163Z DEBUG topgrade::runner   > Step "Brew"
 2020-10-07T09:53:31.163Z DEBUG topgrade::runner   > Step "yadm"
 2020-10-07T09:53:31.163Z DEBUG topgrade::runner   > Step "nix"
 2020-10-07T09:53:31.163Z DEBUG topgrade::runner   > Step "home-manager"
 2020-10-07T09:53:31.163Z DEBUG topgrade::runner   > Step "asdf"
 2020-10-07T09:53:31.164Z DEBUG topgrade::utils    > Path "/home/dthomas/.config/emacs" doesn't exist
 2020-10-07T09:53:31.164Z DEBUG topgrade::utils    > Path "/home/dthomas/.emacs.d" doesn't exist
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.doom.d does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.vim does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.config/nvim does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.zshrc does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.tmux does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.config/fish does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.config/openbox does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.config/bspwm does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.config/i3 does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::steps::git > /home/dthomas/.config/sway does not exists
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "Git repositories"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "zr"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "antibody"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "antigen"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "zplug"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "zinit"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "oh-my-zsh"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "fisher"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "oh-my-fish"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "tmux"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "TLDR"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "pearl"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "SDKMAN!"
 2020-10-07T09:53:31.164Z DEBUG topgrade::utils      > Detected "/usr/bin/bash" as "bash"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "apm"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "rustup"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "choosenim"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "cargo"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "Flutter"
 2020-10-07T09:53:31.164Z DEBUG topgrade::runner     > Step "Go"
 2020-10-07T09:53:31.164Z DEBUG topgrade::utils      > Detected "/usr/bin/go" as "go"
 2020-10-07T09:53:31.164Z DEBUG topgrade::utils      > Path "/home/dthomas/go" exists

―― 20:53:31 - Go ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
 2020-10-07T09:53:31.165Z DEBUG topgrade::executor   > Running "/usr/bin/go" "get" "-u" "all"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "Emacs"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "opam"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "vcpkg"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "pipx"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "stack"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "tlmgr"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "myrepos"
 2020-10-07T09:53:32.350Z DEBUG topgrade::runner     > Step "jetpack"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "vim"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "Neovim"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "voom"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "npm"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "yarn"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "deno"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "composer"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "krew"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "gem"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "sheldon"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "rtcl"
 2020-10-07T09:53:32.351Z DEBUG topgrade::runner     > Step "Flatpak"
 2020-10-07T09:53:32.351Z DEBUG topgrade::utils      > Detected "/usr/bin/flatpak" as "flatpak"

―― 20:53:32 - Flatpak User Packages ――――――――――――――――――――――――――――――――――――――――――――
 2020-10-07T09:53:32.351Z DEBUG topgrade::executor   > Running "/usr/bin/flatpak" "update" "--user" "-y"
Looking for updates…
Nothing to do.
 2020-10-07T09:53:32.448Z DEBUG topgrade::executor   > Running "/usr/bin/flatpak" "update" "--system" "-y"
Looking for updates…
Nothing to do.
 2020-10-07T09:53:38.526Z DEBUG topgrade::runner     > Step "snap"
 2020-10-07T09:53:38.527Z DEBUG topgrade::utils      > Detected "/usr/bin/snap" as "snap"

―― 20:53:38 - snap ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
 2020-10-07T09:53:38.528Z DEBUG topgrade::executor   > Running "/usr/bin/sudo" "/usr/bin/snap" "refresh"
All snaps up to date.
 2020-10-07T09:53:42.738Z DEBUG topgrade::runner     > Step "pihole"
 2020-10-07T09:53:42.738Z DEBUG topgrade::runner     > Step "Firmware upgrades"
 2020-10-07T09:53:42.738Z DEBUG topgrade::utils      > Detected "/usr/bin/fwupdmgr" as "fwupdmgr"
 2020-10-07T09:53:42.740Z TRACE topgrade::executor   > Output of "uname" "-r": Output { status: ExitStatus(ExitStatus(0)), stdout: "5.8.13-300.fc33.x86_64\n", stderr: "" }
 2020-10-07T09:53:42.740Z DEBUG topgrade::steps::os::linux > Uname output: 5.8.13-300.fc33.x86_64

―― 20:53:42 - Firmware upgrades ――――――――――――――――――――――――――――――――――――――――――――――――
 2020-10-07T09:53:42.740Z DEBUG topgrade::executor         > Running "/usr/bin/fwupdmgr" "refresh"
Firmware metadata last refresh: 22 hours ago. Use --force to refresh again.
 2020-10-07T09:53:42.784Z DEBUG topgrade::executor         > Running "/usr/bin/fwupdmgr" "get-updates"
• VERTEX4 has no available firmware updates
No updatable devices
 2020-10-07T09:53:42.823Z DEBUG topgrade::runner           > Step "Restarts"

―― 20:53:42 - Summary ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
System update: OK
etc-update: SKIPPED: Cannot find "etc-update" in PATH
Brew: SKIPPED: Cannot find "brew" in PATH
yadm: SKIPPED: Cannot find "yadm" in PATH
nix: SKIPPED: Cannot find "nix" in PATH
home-manager: SKIPPED: Cannot find "home-manager" in PATH
asdf: SKIPPED: Cannot find "asdf" in PATH
Git repositories: SKIPPED: No repositories to pull
zr: SKIPPED: Cannot find "zsh" in PATH
antibody: SKIPPED: Cannot find "zsh" in PATH
antigen: SKIPPED: Cannot find "zsh" in PATH
zplug: SKIPPED: Cannot find "zsh" in PATH
zinit: SKIPPED: Cannot find "zsh" in PATH
oh-my-zsh: SKIPPED: Cannot find "zsh" in PATH
fisher: SKIPPED: Cannot find "fish" in PATH
oh-my-fish: SKIPPED: Cannot find "fish" in PATH
tmux: SKIPPED: Path "/home/dthomas/.tmux/plugins/tpm/bin/update_plugins" doesn't exist
TLDR: SKIPPED: Cannot find "tldr" in PATH
pearl: SKIPPED: Cannot find "pearl" in PATH
SDKMAN!: SKIPPED: Path "/home/dthomas/.sdkman/bin/sdkman-init.sh" doesn't exist
apm: SKIPPED: Cannot find "apm" in PATH
rustup: SKIPPED: Cannot find "rustup" in PATH
choosenim: SKIPPED: Cannot find "choosenim" in PATH
cargo: SKIPPED: Cannot find "cargo-install-update" in PATH
Flutter: SKIPPED: Cannot find "flutter" in PATH
Go: OK
Emacs: SKIPPED: Cannot find "emacs" in PATH
opam: SKIPPED: Cannot find "opam" in PATH
vcpkg: SKIPPED: Cannot find "vcpkg" in PATH
pipx: SKIPPED: Cannot find "pipx" in PATH
stack: SKIPPED: Cannot find "stack" in PATH
tlmgr: SKIPPED: tlmgr must be explicity enabled in the configuration to run in Linux
myrepos: SKIPPED: Cannot find "mr" in PATH
jetpack: SKIPPED: Cannot find "jetpack" in PATH
vim: SKIPPED: Cannot find "vim" in PATH
Neovim: SKIPPED: Cannot find "nvim" in PATH
voom: SKIPPED: Cannot find "voom" in PATH
npm: SKIPPED: Cannot find "npm" in PATH
yarn: SKIPPED: Cannot find "yarn" in PATH
deno: SKIPPED: Cannot find "deno" in PATH
composer: SKIPPED: Cannot find "composer" in PATH
krew: SKIPPED: Cannot find "kubectl-krew" in PATH
gem: SKIPPED: Cannot find "gem" in PATH
sheldon: SKIPPED: Cannot find "sheldon" in PATH
rtcl: SKIPPED: Cannot find "rupdate" in PATH
Flatpak: OK
snap: OK
pihole: SKIPPED: Cannot find "pihole" in PATH
Firmware upgrades: OK
Restarts: SKIPPED: Cannot find "needrestart" in PATH
</pre></details>

**Comment by r-darwish** [comment]
I don't really know Go well but running `go get` as root sounds dangerous and undesirable. Is it possible that you have some permission problem in your system? What's the permission of the file `/usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/ppc64.a`?

**Comment by donthomas21** [comment]
The whole directory is root:root and the permissions are 644 so no joy there. No real option given the that go was installed with dnf.

I will remove it and try to build manually, that should be more flexible.

**Comment by r-darwish** [comment]
Seems odd. As I understand Go tries to read from that directory, not write. Can you perform `cp /home/dthomas/.cache/go-build/9a/9a1c18c66649a9dc9a60adf0036cbbffcb117a3be8ff8b12a53c8f3aa24dbdbb-d: open /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/ppc64.a /tmp/ppc64.a` as your regular user?

**Comment by donthomas21** [comment]
Copy worked fine, everyone has read permission in /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/

Problem appears to come in the update while trying to copy from .cache to /usr/lib/golang/pkg/linux_amd64/cmd/internal/obj/

**Comment by r-darwish** [comment]
What's the output of `go env`?

**Comment by donthomas21** [comment]
Here is the output from go env

> go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/dthomas/.cache/go-build"
GOENV="/home/dthomas/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/dthomas/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/dthomas/go"
GOPRIVATE=""
GOPROXY="direct"
GOROOT="/usr/lib/golang"
GOSUMDB="off"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/golang/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build340753614=/tmp/go-build -gno-record-gcc-switches"

**Comment by r-darwish** [comment]
Seems reasonable as GOPATH is at your home directory. I'll wait for someone who actually understands go to explain why sudo permissions are needed specifically in fedora.

**Comment by andrei-pavel** [comment]
Found this related to cross-compiling: https://github.com/golang/go/issues/14213
Notice that all your errors are related to different architectures i.e. arm, arm64, mips, ppc64 than your own architecture i.e. amd64 so this isn't a far fetch. The proposed solution in that issue to use `go build` instead isn't helping much though in this situation.
What I would try is to chown $GOROOT to yourself `sudo chown -R "${USER}:$(stat -c %G "${HOME}")" /usr/lib/golang` if you're the only user on the system.

**Comment by r-darwish** [comment]
@andrei-pavel thank you for clarifying. I'll close this issue as this is an issue with Go.

**Comment by andrei-pavel** [comment]
@donthomas21 We could chat here to try to make some progress. I also have Fedora, only it's 32 and I also have go installed through dnf and I also have $GOROOT set to `/usr/lib/golang` and I also have files in there set to root:root 644 and directories to root:root 755, but when I did `go get -u all` I got a different error about `missing packages` which probably has something to do with go ignoring versioning of dependencies. So it seems to have gone further from your point of error.

Since then, I've tried `go get -u all` inside $GOROOT just to see what happens and it seemed to update some stuff, but now wherever I run it from, it just hangs so I seem to have bricked it. It really is a problematic command.

**Comment by donthomas21** [comment]
Hi Andrei,

Thanks for the offer of help with this little problem. Unfortunately, I only have two computers with F33 install and none with F32. The only reason I installed “go” was to compile a command line utility called “duf”

I only included a representative list of the errors so if may not be limited the architectures identified in the errors.

Unfortunately, I ran the ‘go get -u all’ command as sudo and there do not appear to be any subsequent failures in when updating go.

I will redo the installation of F33 as I’m having some unrelated problems (vnc is being annoying, but Fedora just updated it today).

I’ll let you know how the new install goes.

Thanks,

Don
